File: server/engine/geospatial_engine.js
/**
 * Geo-Spatial Disease Mapping Engine
 * Advanced spatial analysis for drone-based disease detection
 */
class GeoSpatialEngine {
    constructor() {
        this.EARTH_RADIUS_M = 6371000;
    }

    // --- Core: Tiling & Heatmap ---

    generateHeatmap(fieldBoundary, scanPoints, options = {}) {
        const cellSize = options.cellSize || 10;

        // 1. Grid Generation
        const tiles = this.createTiles(fieldBoundary, cellSize);

        // 2. Spatial Indexing & Assignment
        const tilesWithData = this.assignScansToTiles(tiles, scanPoints);

        // 3. Scoring
        const scoredTiles = this.computeTileScores(tilesWithData);

        // 4. Clustering (Infected Zones)
        const clusters = this.detectClusters(scoredTiles);

        return {
            heatmap: this.toGeoJSON(scoredTiles),
            zones: this.clustersToGeoJSON(clusters)
        };
    }

    createTiles(fieldBoundary, cellSizeM) {
        const polygon = fieldBoundary.polygon;
        const lats = polygon.map(p => p[0]);
        const lngs = polygon.map(p => p[1]);
        const minLat = Math.min(...lats);
        const maxLat = Math.max(...lats);
        const minLng = Math.min(...lngs);
        const maxLng = Math.max(...lngs);

        // Deg delta approx
        const latDelta = cellSizeM / 111000;
        const avgLat = (minLat + maxLat) / 2;
        const lngDelta = cellSizeM / (111000 * Math.cos(avgLat * Math.PI / 180));

        const tiles = [];
        let tileId = 0;

        for (let lat = minLat; lat < maxLat; lat += latDelta) {
            for (let lng = minLng; lng < maxLng; lng += lngDelta) {
                const center = [lat + latDelta / 2, lng + lngDelta / 2];
                // Simple bounding box check first, then precise PIP
                if (this.pointInPolygon(center, polygon)) {
                    tiles.push({
                        tile_id: tileId++,
                        bounds: [
                            [lat, lng],
                            [lat + latDelta, lng],
                            [lat + latDelta, lng + lngDelta],
                            [lat, lng + lngDelta],
                            [lat, lng]
                        ],
                        center,
                        scans: [],
                        detections: []
                    });
                }
            }
        }
        return tiles;
    }

    assignScansToTiles(tiles, scanPoints) {
        // Optimization: In a real PostGIS DB, this is a spatial join.
        // Here we iterate. For large datasets, use a QuadTree (skipped for simplicity).
        scanPoints.forEach(scan => {
            // Find closest tile or containing tile
            const tile = tiles.find(t => this.isInsideBounds(scan, t.bounds));
            if (tile) {
                tile.scans.push(scan);
                if (scan.detections) tile.detections.push(...scan.detections);
            }
        });
        return tiles;
    }

    isInsideBounds(scan, bounds) {
        // Simple rectangle check for grid tiles (faster than general polygon)
        // bounds: [SW, NW, NE, SE, SW]
        const minLat = bounds[0][0];
        const maxLat = bounds[2][0];
        const minLng = bounds[0][1];
        const maxLng = bounds[2][1];
        return scan.gps_lat >= minLat && scan.gps_lat <= maxLat &&
            scan.gps_lng >= minLng && scan.gps_lng <= maxLng;
    }

    pointInPolygon(point, polygon) {
        const [x, y] = point;
        let inside = false;
        for (let i = 0, j = polygon.length - 1; i < polygon.length; j = i++) {
            const [xi, yi] = polygon[i];
            const [xj, yj] = polygon[j];
            const intersect = ((yi > y) !== (yj > y)) && (x < (xj - xi) * (y - yi) / (yj - yi) + xi);
            if (intersect) inside = !inside;
        }
        return inside;
    }

    computeTileScores(tiles) {
        return tiles.map(tile => {
            if (tile.detections.length === 0) return { ...tile, score: 0, severity: 'none', detection_count: 0 };

            // Metrics
            const count = tile.detections.length;
            const avgConf = tile.detections.reduce((s, d) => s + (d.confidence || 0), 0) / count;
            const avgArea = tile.detections.reduce((s, d) => s + (d.bbox_area_percent || 0), 0) / count;

            // Score Formula (0-100)
            // Weigh area heavily as it indicates severity
            let rawScore = (avgArea * 15) + (count * 5);
            let score = Math.min(Math.round(rawScore * avgConf), 100);

            let severity = score > 60 ? 'high' : score > 30 ? 'moderate' : 'low';

            // Dominant disease
            const counts = {};
            tile.detections.forEach(d => {
                const cls = d.class_name || 'unknown';
                counts[cls] = (counts[cls] || 0) + 1;
            });
            const dominant = Object.keys(counts).sort((a, b) => counts[b] - counts[a])[0];

            return {
                ...tile,
                score,
                severity,
                dominant_disease: dominant,
                detection_count: count,
                avg_area_percent: avgArea
            };
        });
    }

    // --- Clustering (DBSCAN-like) ---

    detectClusters(tiles) {
        const threshold = 30; // Min score to be part of a cluster
        const highRisk = tiles.filter(t => t.score >= threshold);
        const clusters = [];
        const visited = new Set();
        const SEARCH_DIST_M = 20; // Neighbor distance

        highRisk.forEach(tile => {
            if (visited.has(tile.tile_id)) return;

            const cluster = [tile];
            visited.add(tile.tile_id);
            const queue = [tile];

            while (queue.length > 0) {
                const current = queue.shift();

                // Find neighbors in highRisk set
                const neighbors = highRisk.filter(n =>
                    !visited.has(n.tile_id) &&
                    this.distance(current.center, n.center) <= SEARCH_DIST_M
                );

                neighbors.forEach(n => {
                    visited.add(n.tile_id);
                    cluster.push(n);
                    queue.push(n);
                });
            }

            clusters.push(cluster);
        });

        return clusters;
    }

    clustersToGeoJSON(clusters) {
        return {
            type: 'FeatureCollection',
            features: clusters.map((cluster, idx) => {
                const lats = cluster.map(t => t.center[0]);
                const lngs = cluster.map(t => t.center[1]);
                const minLat = Math.min(...lats);
                const maxLat = Math.max(...lats);
                const minLng = Math.min(...lngs);
                const maxLng = Math.max(...lngs);

                // Expand slightly to cover tile width (~0.0001 deg)
                const padding = 0.00005;

                return {
                    type: 'Feature',
                    geometry: {
                        type: 'Polygon',
                        coordinates: [[
                            [minLng - padding, minLat - padding],
                            [maxLng + padding, minLat - padding],
                            [maxLng + padding, maxLat + padding],
                            [minLng - padding, maxLat + padding],
                            [minLng - padding, minLat - padding]
                        ]]
                    },
                    properties: {
                        zone_id: idx + 1,
                        score: Math.round(cluster.reduce((a, b) => a + b.score, 0) / cluster.length),
                        tile_count: cluster.length,
                        disease: cluster[0].dominant_disease,
                        area_m2: cluster.length * 100 // approx 10x10m tiles
                    }
                };
            })
        };
    }

    generateStats(tiles) {
        const totalTiles = tiles.length;
        if (totalTiles === 0) return { infected_area_percent: 0, dominant_disease: 'none', hotspots: [] };

        const infectedTiles = tiles.filter(t => t.score > 20);
        const infectedAreaPct = ((infectedTiles.length / totalTiles) * 100).toFixed(1);

        const diseaseCounts = {};
        infectedTiles.forEach(t => {
            if (t.dominant_disease) diseaseCounts[t.dominant_disease] = (diseaseCounts[t.dominant_disease] || 0) + 1;
        });

        const dominant = Object.keys(diseaseCounts).sort((a, b) => diseaseCounts[b] - diseaseCounts[a])[0] || 'none';

        const hotspots = tiles
            .sort((a, b) => b.score - a.score)
            .slice(0, 5)
            .map(t => ({
                id: t.tile_id,
                score: t.score,
                location: t.center
            }));

        return {
            infected_area_percent: parseFloat(infectedAreaPct),
            dominant_disease: dominant,
            hotspots: hotspots,
            total_area_scanned_m2: totalTiles * 100
        };
    }

    // --- Spread Direction ---

    calculateSpread(scanPoints, diseaseClass = null) {
        // Filter
        let data = scanPoints;
        if (diseaseClass) {
            data = data.filter(s => s.detections.some(d => d.class_name === diseaseClass));
        }

        // Group by Date (YYYY-MM-DD)
        const groups = {};
        data.forEach(s => {
            const date = s.timestamp.split('T')[0];
            if (!groups[date]) groups[date] = { latSum: 0, lngSum: 0, count: 0, timestamp: s.timestamp };
            groups[date].latSum += s.gps_lat;
            groups[date].lngSum += s.gps_lng;
            groups[date].count++;
        });

        // Compute centroids
        const centroids = Object.keys(groups).sort().map(date => {
            const g = groups[date];
            return {
                date,
                lat: g.latSum / g.count,
                lng: g.lngSum / g.count
            };
        });

        if (centroids.length < 2) return null;

        // Compute vectors between consecutive days
        const vectors = [];
        for (let i = 0; i < centroids.length - 1; i++) {
            const curr = centroids[i];
            const next = centroids[i + 1];

            const dist = this.distance([curr.lat, curr.lng], [next.lat, next.lng]);
            const bearing = this.bearing([curr.lat, curr.lng], [next.lat, next.lng]);

            vectors.push({
                from_date: curr.date,
                to_date: next.date,
                start: [curr.lat, curr.lng],
                end: [next.lat, next.lng],
                distance_m: dist,
                bearing_deg: bearing
            });
        }

        return {
            disease: diseaseClass || 'all',
            vectors,
            centroids
        };
    }

    // --- Utils ---

    distance(p1, p2) {
        const R = 6371e3;
        const φ1 = p1[0] * Math.PI / 180;
        const φ2 = p2[0] * Math.PI / 180;
        const Δφ = (p2[0] - p1[0]) * Math.PI / 180;
        const Δλ = (p2[1] - p1[1]) * Math.PI / 180;

        const a = Math.sin(Δφ / 2) * Math.sin(Δφ / 2) +
            Math.cos(φ1) * Math.cos(φ2) *
            Math.sin(Δλ / 2) * Math.sin(Δλ / 2);
        const c = 2 * Math.atan2(Math.sqrt(a), Math.sqrt(1 - a));
        return R * c;
    }

    bearing(p1, p2) {
        const λ1 = p1[1] * Math.PI / 180;
        const λ2 = p2[1] * Math.PI / 180;
        const φ1 = p1[0] * Math.PI / 180;
        const φ2 = p2[0] * Math.PI / 180;

        const y = Math.sin(λ2 - λ1) * Math.cos(φ2);
        const x = Math.cos(φ1) * Math.sin(φ2) -
            Math.sin(φ1) * Math.cos(φ2) * Math.cos(λ2 - λ1);
        const θ = Math.atan2(y, x);
        const brng = (θ * 180 / Math.PI + 360) % 360;
        return brng;
    }

    toGeoJSON(tiles) {
        return {
            type: "FeatureCollection",
            features: tiles.map(t => ({
                type: "Feature",
                geometry: {
                    type: "Polygon",
                    coordinates: [t.bounds.map(p => [p[1], p[0]])] // GeoJSON is Lng,Lat
                },
                properties: {
                    id: t.tile_id,
                    score: t.score,
                    severity: t.severity,
                    disease: t.dominant_disease,
                    count: t.detection_count
                }
            }))
        };
    }

    // --- Advanced Progression Analysis ---

    analyzeProgression(fieldBoundary, allScans, options = {}) {
        const windowDays = options.days || 7;
        const diseaseClass = options.disease || null;

        // 1. Filter by Disease
        let relevantScans = allScans;
        if (diseaseClass && diseaseClass !== 'all') {
            relevantScans = relevantScans.filter(s =>
                s.detections.some(d => d.class_name === diseaseClass)
            );
        }

        // 2. Time Splitting
        const now = new Date();
        const splitPoint = new Date(now.getTime() - windowDays * 86400000);
        const startPoint = new Date(now.getTime() - 2 * windowDays * 86400000);

        const currentScans = relevantScans.filter(s => new Date(s.timestamp) >= splitPoint);
        const previousScans = relevantScans.filter(s => {
            const d = new Date(s.timestamp);
            return d >= startPoint && d < splitPoint;
        });

        // 3. Generate Maps for Comparison
        const gridOpts = { cellSize: 10 };
        const currMap = this.generateHeatmap(fieldBoundary, currentScans, gridOpts);
        const prevMap = this.generateHeatmap(fieldBoundary, previousScans, gridOpts);

        // 4. Calculate Centroids
        const getCentroid = (features) => {
            let sumLat = 0, sumLng = 0, totalScore = 0;
            features.forEach(f => {
                if (f.properties.score > 20) { // Only consider active infection
                    const [lng, lat] = f.geometry.type === 'Polygon'
                        ? f.geometry.coordinates[0][0] // simplistic center
                        : f.geometry.coordinates; // Point

                    const w = f.properties.score;
                    sumLat += lat * w;
                    sumLng += lng * w;
                    totalScore += w;
                }
            });
            return totalScore > 0 ? [sumLat / totalScore, sumLng / totalScore] : null;
        };

        const c1 = getCentroid(currMap.heatmap.features);
        const c2 = getCentroid(prevMap.heatmap.features);

        // 5. Compute Metrics
        const getInfectedArea = (features) => features.filter(f => f.properties.score > 20).length * 100; // m2
        const currArea = getInfectedArea(currMap.heatmap.features);
        const prevArea = getInfectedArea(prevMap.heatmap.features);

        const deltaArea = currArea - prevArea;
        const status = deltaArea > 0 ? "EXPANDING" : deltaArea < 0 ? "SHRINKING" : "STABLE";

        // 6. Identify New Hotspots (Tiles newly infected)
        const prevInfectedIds = new Set(
            prevMap.heatmap.features
                .filter(f => f.properties.score > 20)
                .map(f => f.properties.id)
        );

        const newInfections = currMap.heatmap.features
            .filter(f => f.properties.score > 20 && !prevInfectedIds.has(f.properties.id))
            .map(f => ({
                tile_id: f.properties.id,
                score: f.properties.score,
                coordinates: f.geometry.coordinates[0][0] // approx
            }));

        // 7. Spread Vector (GeoJSON Arrow)
        let spreadVector = null;
        if (c1 && c2) {
            const dist = this.distance(c2, c1); // prev -> curr
            const bearing = this.bearing(c2, c1);
            spreadVector = {
                type: "Feature",
                geometry: {
                    type: "LineString",
                    coordinates: [[c2[1], c2[0]], [c1[1], c1[0]]] // GeoJSON Lng,Lat
                },
                properties: {
                    from_date: startPoint.toISOString().split('T')[0],
                    to_date: now.toISOString().split('T')[0],
                    distance_m: Math.round(dist),
                    bearing_deg: Math.round(bearing),
                    direction: this.getCardinalDirection(bearing)
                }
            };
        }

        return {
            window_days: windowDays,
            disease: diseaseClass || 'all',
            status,
            metrics: {
                current_infected_area_m2: currArea,
                previous_infected_area_m2: prevArea,
                area_change_m2: deltaArea,
                growth_pct: prevArea > 0 ? Math.round((deltaArea / prevArea) * 100) : 100
            },
            spread_vector: spreadVector,
            centroid_shift: {
                previous: c2,
                current: c1
            },
            new_hotspots: newInfections,
            timeline: [
                { date: startPoint.toISOString().split('T')[0], area: prevArea },
                { date: splitPoint.toISOString().split('T')[0], area: currArea }
            ]
        };
    }

    getCardinalDirection(deg) {
        const dirs = ['N', 'NE', 'E', 'SE', 'S', 'SW', 'W', 'NW'];
        return dirs[Math.round(deg / 45) % 8];
    }
}

module.exports = new GeoSpatialEngine();

File: server/engine/geospatial_engine.py
import math
import datetime

class GeoSpatialEngine:
    def __init__(self):
        self.EARTH_RADIUS_M = 6371000

    def generate_heatmap(self, field_boundary, scan_points, options=None):
        options = options or {}
        cell_size = options.get("cellSize", 10)

        tiles = self.create_tiles(field_boundary, cell_size)
        tiles_with_data = self.assign_scans_to_tiles(tiles, scan_points)
        scored_tiles = self.compute_tile_scores(tiles_with_data)
        clusters = self.detect_clusters(scored_tiles)

        return {
            "heatmap": self.to_geojson(scored_tiles),
            "zones": self.clusters_to_geojson(clusters)
        }

    def create_tiles(self, field_boundary, cell_size_m):
        polygon = field_boundary.get("polygon", [])
        if not polygon: return []
        
        lats = [p[0] for p in polygon]
        lngs = [p[1] for p in polygon]
        
        min_lat, max_lat = min(lats), max(lats)
        min_lng, max_lng = min(lngs), max(lngs)

        lat_delta = cell_size_m / 111000.0
        avg_lat = (min_lat + max_lat) / 2
        lng_delta = cell_size_m / (111000.0 * math.cos(avg_lat * math.pi / 180))

        tiles = []
        tile_id = 0

        lat = min_lat
        while lat < max_lat:
            lng = min_lng
            while lng < max_lng:
                center = [lat + lat_delta / 2, lng + lng_delta / 2]
                if self.point_in_polygon(center, polygon):
                    tiles.append({
                        "tile_id": tile_id,
                        "bounds": [
                            [lat, lng],
                            [lat + lat_delta, lng],
                            [lat + lat_delta, lng + lng_delta],
                            [lat, lng + lng_delta],
                            [lat, lng]
                        ],
                        "center": center,
                        "scans": [],
                        "detections": []
                    })
                    tile_id += 1
                lng += lng_delta
            lat += lat_delta
            
        return tiles

    def assign_scans_to_tiles(self, tiles, scan_points):
        for scan in scan_points:
            for tile in tiles:
                if self.is_inside_bounds(scan, tile["bounds"]):
                    tile["scans"].append(scan)
                    if scan.get("detections"):
                        tile["detections"].extend(scan["detections"])
                    break
        return tiles

    def is_inside_bounds(self, scan, bounds):
        min_lat, max_lat = bounds[0][0], bounds[2][0]
        min_lng, max_lng = bounds[0][1], bounds[2][1]
        lat = scan.get("gps_lat", 0)
        lng = scan.get("gps_lng", 0)
        return min_lat <= lat <= max_lat and min_lng <= lng <= max_lng

    def point_in_polygon(self, point, polygon):
        x, y = point
        inside = False
        j = len(polygon) - 1
        for i in range(len(polygon)):
            xi, yi = polygon[i]
            xj, yj = polygon[j]
            intersect = ((yi > y) != (yj > y)) and (x < (xj - xi) * (y - yi) / (yj - yi) + xi)
            if intersect:
                inside = not inside
            j = i
        return inside

    def compute_tile_scores(self, tiles):
        for tile in tiles:
            dets = tile["detections"]
            if not dets:
                tile.update({"score": 0, "severity": "none", "detection_count": 0})
                continue

            count = len(dets)
            avg_conf = sum(d.get("confidence", 0) for d in dets) / count
            avg_area = sum(d.get("bbox_area_percent", 0) for d in dets) / count

            raw_score = (avg_area * 15) + (count * 5)
            score = min(round(raw_score * avg_conf), 100)

            severity = "high" if score > 60 else "moderate" if score > 30 else "low"

            counts = {}
            for d in dets:
                cls = d.get("class_name") or "unknown"
                counts[cls] = counts.get(cls, 0) + 1
            
            dominant = max(counts, key=counts.get) if counts else "none"

            tile.update({
                "score": score,
                "severity": severity,
                "dominant_disease": dominant,
                "detection_count": count,
                "avg_area_percent": avg_area
            })
        return tiles

    def detect_clusters(self, tiles):
        threshold = 30
        high_risk = [t for t in tiles if t.get("score", 0) >= threshold]
        
        clusters = []
        visited = set()
        search_dist_m = 20

        for tile in high_risk:
            tid = tile["tile_id"]
            if tid in visited: continue
            
            cluster = [tile]
            visited.add(tid)
            queue = [tile]
            
            while queue:
                current = queue.pop(0)
                neighbors = [n for n in high_risk if n["tile_id"] not in visited and self.distance(current["center"], n["center"]) <= search_dist_m]
                
                for n in neighbors:
                    visited.add(n["tile_id"])
                    cluster.append(n)
                    queue.append(n)
                    
            clusters.append(cluster)
            
        return clusters

    def clusters_to_geojson(self, clusters):
        features = []
        for idx, cluster in enumerate(clusters):
            lats = [t["center"][0] for t in cluster]
            lngs = [t["center"][1] for t in cluster]
            min_lat, max_lat = min(lats), max(lats)
            min_lng, max_lng = min(lngs), max(lngs)
            
            padding = 0.00005
            
            score = round(sum(t.get("score", 0) for t in cluster) / len(cluster)) if cluster else 0
            
            features.append({
                "type": "Feature",
                "geometry": {
                    "type": "Polygon",
                    "coordinates": [[
                        [min_lng - padding, min_lat - padding],
                        [max_lng + padding, min_lat - padding],
                        [max_lng + padding, max_lat + padding],
                        [min_lng - padding, max_lat + padding],
                        [min_lng - padding, min_lat - padding]
                    ]]
                },
                "properties": {
                    "zone_id": idx + 1,
                    "score": score,
                    "tile_count": len(cluster),
                    "disease": cluster[0].get("dominant_disease") if cluster else "none",
                    "area_m2": len(cluster) * 100
                }
            })
            
        return {"type": "FeatureCollection", "features": features}

    def generate_stats(self, tiles):
        total_tiles = len(tiles)
        if total_tiles == 0:
            return {"infected_area_percent": 0.0, "dominant_disease": "none", "hotspots": [], "total_area_scanned_m2": 0}
            
        infected = [t for t in tiles if t.get("score", 0) > 20]
        infected_pct = round((len(infected) / total_tiles) * 100, 1)
        
        counts = {}
        for t in infected:
            d = t.get("dominant_disease")
            if d: counts[d] = counts.get(d, 0) + 1
            
        dominant = max(counts, key=counts.get) if counts else "none"
        
        sorted_tiles = sorted(tiles, key=lambda x: x.get("score", 0), reverse=True)[:5]
        hotspots = [{"id": t.get("tile_id"), "score": t.get("score"), "location": t.get("center")} for t in sorted_tiles if t.get("score", 0) > 0]
        
        return {
            "infected_area_percent": infected_pct,
            "dominant_disease": dominant,
            "hotspots": hotspots,
            "total_area_scanned_m2": total_tiles * 100
        }

    def calculate_spread(self, scan_points, disease_class=None):
        data = scan_points
        if disease_class:
            data = [s for s in data if any(d.get("class_name") == disease_class for d in s.get("detections", []))]
            
        groups = {}
        for s in data:
            date_str = str(s.get("timestamp", ""))[:10]
            if date_str not in groups:
                groups[date_str] = {"latSum": 0, "lngSum": 0, "count": 0}
            groups[date_str]["latSum"] += s.get("gps_lat", 0)
            groups[date_str]["lngSum"] += s.get("gps_lng", 0)
            groups[date_str]["count"] += 1
            
        centroids = []
        for date_str in sorted(groups.keys()):
            g = groups[date_str]
            if g["count"] > 0:
                centroids.append({
                    "date": date_str,
                    "lat": g["latSum"] / g["count"],
                    "lng": g["lngSum"] / g["count"]
                })
                
        if len(centroids) < 2: return None
        
        vectors = []
        for i in range(len(centroids) - 1):
            curr = centroids[i]
            nxt = centroids[i+1]
            dist = self.distance([curr["lat"], curr["lng"]], [nxt["lat"], nxt["lng"]])
            bearing = self.bearing([curr["lat"], curr["lng"]], [nxt["lat"], nxt["lng"]])
            vectors.append({
                "from_date": curr["date"],
                "to_date": nxt["date"],
                "start": [curr["lat"], curr["lng"]],
                "end": [nxt["lat"], nxt["lng"]],
                "distance_m": dist,
                "bearing_deg": bearing
            })
            
        return {"disease": disease_class or "all", "vectors": vectors, "centroids": centroids}

    def analyze_progression(self, field_boundary, all_scans, options=None):
        options = options or {}
        window_days = options.get("days", 7)
        disease_class = options.get("disease", None)
        
        relevant_scans = all_scans
        if disease_class and disease_class != "all":
            relevant_scans = [s for s in relevant_scans if any(d.get("class_name") == disease_class for d in s.get("detections", []))]
            
        now = datetime.datetime.now()
        split_point = now - datetime.timedelta(days=window_days)
        start_point = now - datetime.timedelta(days=window_days * 2)
        
        curr_scans = []
        prev_scans = []
        
        for s in relevant_scans:
            try:
                date_val = datetime.datetime.fromisoformat(s.get("timestamp", "2000-01-01").replace("Z", "+00:00"))
                # Using timezone unaware to be simple
                date_val = date_val.replace(tzinfo=None)
                if date_val >= split_point:
                    curr_scans.append(s)
                elif start_point <= date_val < split_point:
                    prev_scans.append(s)
            except:
                curr_scans.append(s)
                
        curr_map = self.generate_heatmap(field_boundary, curr_scans, {"cellSize": 10})
        prev_map = self.generate_heatmap(field_boundary, prev_scans, {"cellSize": 10})
        
        def get_centroid(features):
            sum_lat, sum_lng, total_score = 0, 0, 0
            for f in features:
                score = f["properties"].get("score", 0)
                if score > 20:
                    geom = f["geometry"]["coordinates"]
                    lng, lat = geom[0][0][0], geom[0][0][1]
                    sum_lat += lat * score
                    sum_lng += lng * score
                    total_score += score
            return [sum_lat / total_score, sum_lng / total_score] if total_score > 0 else None
            
        c1 = get_centroid(curr_map["heatmap"]["features"])
        c2 = get_centroid(prev_map["heatmap"]["features"])
        
        def get_infected_area(features):
            return sum(1 for f in features if f["properties"].get("score", 0) > 20) * 100
            
        curr_area = get_infected_area(curr_map["heatmap"]["features"])
        prev_area = get_infected_area(prev_map["heatmap"]["features"])
        
        delta_area = curr_area - prev_area
        status = "EXPANDING" if delta_area > 0 else "SHRINKING" if delta_area < 0 else "STABLE"
        
        prev_infected_ids = {f["properties"].get("id") for f in prev_map["heatmap"]["features"] if f["properties"].get("score", 0) > 20}
        
        new_infections = []
        for f in curr_map["heatmap"]["features"]:
            if f["properties"].get("score", 0) > 20 and f["properties"].get("id") not in prev_infected_ids:
                new_infections.append({
                    "tile_id": f["properties"].get("id"),
                    "score": f["properties"].get("score"),
                    "coordinates": f["geometry"]["coordinates"][0][0]
                })
                
        spread_vector = None
        if c1 and c2:
            dist = self.distance(c2, c1)
            bearing = self.bearing(c2, c1)
            spread_vector = {
                "type": "Feature",
                "geometry": {
                    "type": "LineString",
                    "coordinates": [[c2[1], c2[0]], [c1[1], c1[0]]]
                },
                "properties": {
                    "from_date": start_point.strftime("%Y-%m-%d"),
                    "to_date": now.strftime("%Y-%m-%d"),
                    "distance_m": round(dist),
                    "bearing_deg": round(bearing),
                    "direction": self.get_cardinal_direction(bearing)
                }
            }
            
        return {
            "window_days": window_days,
            "disease": disease_class or "all",
            "status": status,
            "metrics": {
                "current_infected_area_m2": curr_area,
                "previous_infected_area_m2": prev_area,
                "area_change_m2": delta_area,
                "growth_pct": round((delta_area / prev_area) * 100) if prev_area > 0 else 100
            },
            "spread_vector": spread_vector,
            "centroid_shift": {"previous": c2, "current": c1},
            "new_hotspots": new_infections,
            "timeline": [
                {"date": start_point.strftime("%Y-%m-%d"), "area": prev_area},
                {"date": split_point.strftime("%Y-%m-%d"), "area": curr_area}
            ]
        }

    def distance(self, p1, p2):
        lat1, lon1 = p1
        lat2, lon2 = p2
        dlat = (lat2 - lat1) * math.pi / 180
        dlon = (lon2 - lon1) * math.pi / 180
        a = math.sin(dlat/2)**2 + math.cos(lat1*math.pi/180) * math.cos(lat2*math.pi/180) * math.sin(dlon/2)**2
        c = 2 * math.atan2(math.sqrt(a), math.sqrt(1-a))
        return self.EARTH_RADIUS_M * c
        
    def bearing(self, p1, p2):
        lat1, lon1 = p1[0]*math.pi/180, p1[1]*math.pi/180
        lat2, lon2 = p2[0]*math.pi/180, p2[1]*math.pi/180
        dlon = lon2 - lon1
        y = math.sin(dlon) * math.cos(lat2)
        x = math.cos(lat1) * math.sin(lat2) - math.sin(lat1) * math.cos(lat2) * math.cos(dlon)
        brng = math.atan2(y, x) * 180 / math.pi
        return (brng + 360) % 360

    def get_cardinal_direction(self, deg):
        dirs = ['N', 'NE', 'E', 'SE', 'S', 'SW', 'W', 'NW']
        return dirs[int(round(deg / 45)) % 8]

    def to_geojson(self, tiles):
        features = [{
            "type": "Feature",
            "geometry": {
                "type": "Polygon",
                "coordinates": [[ [p[1], p[0]] for p in t["bounds"] ]]
            },
            "properties": {
                "id": t.get("tile_id"),
                "score": t.get("score"),
                "severity": t.get("severity"),
                "disease": t.get("dominant_disease"),
                "count": t.get("detection_count")
            }
        } for t in tiles]
        return {"type": "FeatureCollection", "features": features}

geo_spatial_engine = GeoSpatialEngine()
